add development instructions for api endpoints (#70)
BobaBoard/boba-docs: docusaurus/docs/engineering/boba-backend/backend-development-outline.md

---
sidebar_position: 2
---

# Backend development overview

While every feature is unique, developing on boba backend usually involves changes to one or more of these entities:

- **API endpoints:** the actions the backend makes available to clients.
- **Permissions:** the flags that can be associated to a role to allow users to take specific actions.
- **Database:** the tables and queries used to store and retrieve data from permanent storage.
- **Tests:** the automated checks that ensure code runs (and continues to run) as expected.

> This page goes over the main steps involved in making changes to each of these. This is not the only order

Reviewer: ```suggestion
This guide explains the main steps required to make changes to each of these. This is not the only order
```

- **HTTP method and URL (required):** for example, `DELETE threads/:thread_id`

- **HTTP request payload (optional):** the shape of the data sent by the client as part of the request body (most often with `POST` requests).

  Example (from `POST posts/:post_id/contributions`):

  ```ts
  - content (required): string
  - whisper_tags (optional): string[]
  - index_tags (optional): string[]
  - ...
  ```

- **HTTP response codes and payloads (required):** a list of the response codes returned by the endpoint.

  Example (from `POST posts/:post_id/contributions`):

  - **HTTP 200: Contribution created**

    **Payload:** the contribution just created.

    ```ts
      - content: string
      - whisper_tags: string[]
      - index_tags: string[]
      - ...
    ```

  - **HTTP 401: No authenticated user.**
  - **HTTP 403: Unauthorized user.**
  - **HTTP 404: Post not found.**

### Development phase

During the development phase you should:

1. **Identify where the API endpoint should be placed:** see [Routes and Endpoints Structure](docs/engineering/boba-backend/APIs/creating-endpoints#structure)

> 2. **Write the endpoint function** (if not already present): at this stage, it will simply throw [HTTP 501 Not Implemented](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) error.

Reviewer: ```suggestion
2. **Write the endpoint function** (if not already present): at this stage, it will simply throw and [HTTP 501 Not Implemented](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) error.
```

3. **Add or modify the endpoint OpenAPI definition:** see [Documenting endpoints](docs/engineering/boba-backend/APIs/creating-endpoints#documentation).

   :::TODO

   We'll need to figure out how to mark not implemented functions in our OpenAPI spec.

   :::

4. **Generate Zod and TypeScript definition from OpenAPI (if changed):** Run `yarn run open-api:typescript`, and check that the generated types match your expectations.

   :::TODO

   Document this better and then link to the correct page.

   :::

> 5. **Create a test file:** if no test exists for this endpoint, add it, and test that calling the endpoint correctly returns HTTP 501 Error.

Reviewer: ```suggestion
5. **Create a test file:** if no test exists for this endpoint, add it, and test that calling the endpoint correctly returns an HTTP 501 Error.
```

:::warning

> At this stage, endpoint development might be blocked while needed permissions or database queries are created. You can move on to the appropriate sections are needed, and come back once you're unblocked.

Reviewer: ```suggestion
At this stage, endpoint development might be blocked while needed permissions or database queries are created. You can move on to the appropriate sections as needed, and come back once you're unblocked.
```

:::

> 6. **Add permissions middleware and update tests:** if the endpoint needs to be gated on special conditions (e.g. the user being logged in,

Reviewer: ```suggestion
6. **Add permissions middleware and update tests:** if the endpoint needs to be gated on special conditions (e.g. the user being logged in
```

>    or a permission existing), add the necessary middleware to the endpoint. See [Checking permissions](docs/engineering/boba-backend/permissions/check-permissions). <u>Make sure to also update tests</u> to check the correct HTTP status code are returned.

Reviewer: ```suggestion
   or a permission existing), add the necessary middleware to the endpoint. See [Checking permissions](docs/engineering/boba-backend/permissions/check-permissions). <u>Make sure to also update tests</u> to check the correct HTTP status codes are returned.
```

> 7. **Use SQL queries and generated Zod types to write and validate the endpoint functionality.** With permission errors are taken care of in step 6,

Reviewer: ```suggestion
7. **Use SQL queries and generated Zod types to write and validate the endpoint functionality.** With permission errors taken care of in step 6,
```
